### PR TITLE
[InstCombine] Add fold for fabs(-x) -> fabs(x) 

### PR DESCRIPTION
**Pull request by VaibhavRumale**
Here is my patch for #94170
 
Added the missing optimization in `visitCallInst` function to fold `fabs(-x)` into `fabs(x)'. I also added the required test cases to verify the optimization for various data types, including scalar floats, doubles, and vector types.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 1]
```diff
@@ -2517,6 +2517,13 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
   }
   case Intrinsic::fabs: {
     Value *Cond, *TVal, *FVal;
+    Value* Arg = II->getArgOperand(0);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:2520]: Can you replace all following uses of `II->getArgOperand(0)` with `Arg`?

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 2]
```diff
@@ -2517,6 +2517,13 @@ Instruction *InstCombinerImpl::visitCallInst(CallInst &CI) {
...
+    Value* X;
+    // fabs (-X) --> fabs (X)
+    if (match(Arg, m_FNeg(m_Value(X)))) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:2523]: Please drop the braces.
See https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

File: llvm/test/Transforms/InstCombine/fabs-fneg-fold.ll [hunk 3]
```diff
@@ -0,0 +1,57 @@
+; RUN: opt -S -passes=instcombine %s | FileCheck %s
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/fabs-fneg-fold.ll:1]: Please regenerate check lines with `llvm/utils/update_test_checks.py`

File: llvm/test/Transforms/InstCombine/fabs-fneg-fold.ll [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+; CHECK-LABEL: define half @fabs_fneg_f16(
+; CHECK-NEXT:    [[FABS:%.*]] = call half @llvm.fabs.f16(half [[X:%.*]])
+; CHECK-NEXT:    ret half [[FABS]]
+;
+  %neg = fneg half %x
+  %fabs = call half @llvm.fabs.f16(half %neg)
+  ret half %fabs
+}
+
+declare float @llvm.fabs.f32(float)
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/fabs-fneg-fold.ll:53]: These declarations are unnecessary.

File: llvm/test/Transforms/InstCombine/fabs-fneg-fold.ll [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+define half @fabs_fneg_f16(half %x) {
+; CHECK-LABEL: define half @fabs_fneg_f16(
+; CHECK-NEXT:    [[FABS:%.*]] = call half @llvm.fabs.f16(half [[X:%.*]])
+; CHECK-NEXT:    ret half [[FABS]]
+;
+  %neg = fneg half %x
+  %fabs = call half @llvm.fabs.f16(half %neg)
+  ret half %fabs
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fabs-fneg-fold.ll:52]: Should test flag preservation behavior

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp (hunk 2 above)
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:2523]: Somewhere is a strip sign operations helper function that also covers copysign

**Comment by dtcxzyw** [comment]
@VaibhavRumale Please don't update your branch if there is no merge conflict. See https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

**Comment by dtcxzyw** [comment]
@tstellar Should we turn off `Always suggest updating pull request branches`? I think it is misleading to new contributors.

**Comment by VaibhavRumale** [comment]
@dtcxzyw Apologies! I will be more careful going forward. Thank you!

**Review by arsenm — CHANGES_REQUESTED** [review]
Should test flag preservation behavior. Should also add similar copysign tests, since those should be handled just like fabs

**Comment by VaibhavRumale** [comment]
> @VaibhavRumale Please don't update your branch if there is no merge conflict. See https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

Thanks for sharing the guide. It seems like I made a mistake of committing from my main branch. Can I create a new PR with my new changes, as I know we cannot use forced commits?

**Comment by arsenm** [comment]
> Can I create a new PR with my new changes, as I know we cannot use forced commits?

It's not that you can't, it's just generally discouraged but not always. But you can always create a new one

**Comment by VaibhavRumale** [comment]
Thanks for the reviews. I have opened a new PR #95627 with the changes.

**Pull request closed without merging**